Refit/SystemTextJsonContentSerializer.cs

```diff
@@ -60,13 +60,31 @@ public async Task<T> DeserializeAsync<T>(HttpContent content)
         {
             using var stream = await content.ReadAsStreamAsync().ConfigureAwait(false);
 
-            var buffer = ArrayPool<byte>.Shared.Rent((int)stream.Length);
+            int streamLength;
 
             try
             {
-                var length = await stream.ReadAsync(buffer, 0, buffer.Length).ConfigureAwait(false);
+                streamLength = (int)stream.Length;
```

Comment:
Is there any way we can do this without throwing an exception?

Reply:
It looks like [`HttpBaseStream`](https://github.com/dotnet/runtime/blob/535d957421acdf4bff5e8799d14eaf038ae2fa17/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpBaseStream.cs#L12) is internal, so we couldn't do a `stream is HttpBaseStream` check either.

The only other thing I can think of is to check the type name of the stream in some way, though that's a bit bleurgh and brittle.

Reply:
Yeah that'd be better, agreed. As I've mentioned though, for some reason the `CanSeek` property doesn't always match that behavior, even though technically (according to the docs), it should. My rationale here in writing that workaround was was:

- A `try/catch` block (as far as I know) doesn't add (much) overhead _if an exception is not thrown_. Since I'd expect that to be the most common case, it should be fine.
- If that fails, the additional overhead should still not matter much, as in that branch the code would already fallback to the slower API using an input `Stream`, which would hurt performance on its own compared to the faster branch with the pooled array.

I'm open to suggestions here, not sure either solution is "perfect" 🤔

Reply:
Looking through the code more, it looks like, in this case at least, `CanSeek` is `false`, so the access on `Length` could be guarded by `if (CanSeek)`?

https://github.com/dotnet/runtime/blob/bd192ae891655adcd0e751e9ff4209f4e1bd3221/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpBaseStream.cs#L14

The exception could also maybe by filtered like `catch (NotSupportedException) when (stream.CanSeek)` so it isn't unconditionally swallowing the exception?

Are there scenarios where that wouldn't work? It might be that to ensure the method reasonably always succeeds, we might have to eat the performance benefit and go through the slow path if `Length` isn't guaranteed to work in all cases.

For what it's worth, the slow path for System.Text.Json seems to still be more efficient that Newtonsoft.Json anyway even if it's not _as fast_ as with the pooling.

Reply:
@martincostello Accessing `Length` inside an `if (CanSeek)` check was the original implementation I tried, but then I noticed that as you said, `CanSeek` was `false` even when the `Length` property was actually accessed just fine. The main point is that a `try` block itself doesn't add overhead, in fact it's added implicitly by the C# compiler in each `using` block, which is used pretty often even in high performance code (and it's used even in that very same `DeserializeAsync<T>` method anyway, when getting the input stream. The slowdown would _only_ be when catching the exception, but then again, we don't have a reliable way to avoid that since the `CanSeek` property can sometimes report an incorrect value.

Same goes for a filtered catch, since we're only retrieving the length in that `try/finally` block, regardless of the underlying reason for that exception, if we fail to get that it means that at this point we _have to_ fallback to the `Stream` overload.

With all these things considered, I'm not really seeing why we should just prevent the fast path from ever happening, considering that at least from what I can see here, it's being taken and working properly pretty often (eg. it's used in all the current unit tests).

Note that the point of `System.Text.Json` is not so much the speed improvement (which was about 30% in my tests), but the reduced memory usage (which was pretty big in my tests, [see here](https://github.com/reactiveui/refit/pull/836#issuecomment-583865387)).

Personally I'd prefer to keep the current solution/workaround, at least for now.

Reply:
I’m happy with the PR as it is, but that’s not my call. Was just spit-balling on a possible alternative.

However without a fix of _some kind_ I’ll have to go back to using a custom version rather than this new built-in one.

Reply:
Yeah absolutely, I wasn't arguing with you here, just trying to provide as much feedback and details as possible both for you and for future reference on this PR in particular 😊

I'll see if I can put together some tests so we can move forward with this and hopefully unblock you in your project!